### PR TITLE
Moving serial settings to a separate crate

### LINKED ISSUE #808: Move serial port configuration menu to separate crate
The USB serial configuration menu is a useful utility for serial-based settings configuration used in multiple QUARTIQ firmwares. We should make it generic and have it exist as its own separate crate so we can leverage the functionality for Booster and Thermostat.

### PR DESCRIPTION
**Pull request by ryan-summers**
This PR fixes #808 by creating a new crate for serial settings and storage.

From the serial console:
```
Input was empty?

> help
AVAILABLE ITEMS:
  dfu
  service
  reboot
  factory-reset
  list
  get <item>
  set <item> <value>
  help [ <command> ]

> dfu
Reset to DFU is not supported on this device

> service
Service data not available

> list
Available items:
/broker: "test" [default: "mqtt"]
/id: "04-91-62-d2-a8-6f" [default: "04-91-62-d2-a8-6f"]
```

### CODE REVIEW

File: serial-settings/src/lib.rs
Comment (ryan-summers) [inline serial-settings/src/lib.rs:236]: I suspect we can manage buffering without bbqueue, since there's synchronous access to the various objects.

File: serial-settings/src/lib.rs
Comment (ryan-summers) [inline serial-settings/src/lib.rs:30]: I wonder if there's an easy way for us to make the menu options expandable by the user... Perhaps defer to the future

Comment (jordens) [reply serial-settings/src/lib.rs:30]: Maybe not needed now. All we want to add now is `service` and `dfu` AFAICT.

File: serial-settings/src/lib.rs
Comment (ryan-summers) [inline serial-settings/src/lib.rs:238]: Need to figure out how to load settings from flash and fall-back if failing. In stabilizer, we modify the read-back settings, and I'm not sure the best way to handle that here.

File: serial-settings/src/lib.rs
Comment (ryan-summers) [inline serial-settings/src/lib.rs:286]: I think it may be beneficial to pull out any concept of USB and operate on `Read` / `Write` traits instead, as this would allow this to function over i.e. serial UART as well.

Comment (jordens) [reply serial-settings/src/lib.rs:286]: exactly.

File: serial-settings/src/lib.rs
Comment (ryan-summers) [inline serial-settings/src/lib.rs:18]: No reason to erase all of the flash bank, just erase the area we're going to use.

File: serial-settings/src/lib.rs
Comment (jordens) [reply serial-settings/src/lib.rs:238]: be generic over `<S: Tree + ..., F: FnOnce() -> S>` with `F` providing the default fallback `S`.

File: serial-settings/src/lib.rs
Comment (jordens) [reply serial-settings/src/lib.rs:18]: Oh yes.

File: serial-settings/src/lib.rs
Comment (ryan-summers) [reply serial-settings/src/lib.rs]: Both `service` and `dfu` are not things we have on stabilizer. We can add them, but resetting to DFU mode is generally an involved process, so I'd imagine it's going to be a function provided on the trait object.

File: serial-settings/Cargo.toml
Comment (ryan-summers) [inline serial-settings/Cargo.toml:11]: Currently need https://github.com/rust-embedded-community/menu/pull/15 to make this all work

File: serial-settings/src/lib.rs
Comment (ryan-summers) [reply serial-settings/src/lib.rs]: I added an abstraction in214d29fce5cf46e5cbbd317e5ac0e0051034937b, but I'm not the biggest fan of it because it means that there's `service` and `dfu` menu options in stabilizer that don't do anything currently (although we could add them).

Ideally, these items wouldn't even exist in the menu if they're not desired, but we can hold off on this for the future

**Comment by ryan-summers** [comment]
@jordens I'm curious to see if you have any thoughts with the current design. It's definitely not perfect, but I think it's at a decent point for now. I'm going to poke at integrating it into Booster to see how it goes.

File: serial-settings/src/lib.rs
Comment (ryan-summers) [inline serial-settings/src/lib.rs:60]: This implies that the settings will be persisted to flash, but it's also perfectly valid to persist them to something else, like EEPROM. Ultimately, this just needs to be some non-volatile memory location.

Actually, thinking about it, it doesn't even have to be non-volatile. Should we look for a different interface to use for this?

Comment (ryan-summers) [reply serial-settings/src/lib.rs:60]: I think https://docs.rs/embedded-storage/latest/embedded_storage/trait.Storage.html is the correct trait bound here instead.

Comment (jordens) [reply serial-settings/src/lib.rs:60]: What about moving `save()` into the `Settings` trait and remove it completely from `Context`?

Comment (jordens) [reply serial-settings/src/lib.rs:60]: Or move it into `Platform`

Comment (jordens) [reply serial-settings/src/lib.rs:60]: Also maybe move `Settings` and `Interface` into aassociated types of `Platform`?

Comment (ryan-summers) [reply serial-settings/src/lib.rs]: I moved the various types into associated types of `Platform`, but I haven't touched `save()`.

I'm trying to understand why we would want to move it. Right now, it's abstracting away the concept of saving settings from the user. If we made it an associated function of `Platform`, that means the user could override it with some custom saving method. However, how does that make sense if we're internally loading the settings from the provided `storage` mechanism during initialization? If they were to save it differently, we couldn't load those settings.

It seems like what's currently implemented should be fine for now.

File: serial-settings/src/lib.rs
Comment (jordens) [inline serial-settings/src/lib.rs:229]: I'm a bit unhappy about writing to `context` vs writing to `context.platform.interface_mut()`. They are significantly different.

Comment (jordens) [reply serial-settings/src/lib.rs:229]: Also we should check that the worst case write doesn't unduly starve the other things happening at idle prio (network, streaming IIRC).

File: serial-settings/src/lib.rs
Comment (ryan-summers) [inline serial-settings/src/lib.rs:91]: I'm not following why `P::Settings` aren't just a member of `Context`. It seems like extra work for the implementor of `platform` to have to provide these when in reality the settings are internally managed by this library.

File: serial-settings/src/lib.rs
Comment (ryan-summers) [inline serial-settings/src/lib.rs:79]: Seems like these should accept/return a `Self::Settings` instead of the buffer

File: serial-settings/src/lib.rs
Comment (ryan-summers) [inline serial-settings/src/lib.rs:85]: Similarly, it feels odd that the interface is owned by the underlying `Platform` implementor instead of by the `Context`. The menu is intended to be the primary owner of the interface for communication (at least in my mind).

**Review by ryan-summers — COMMENTED** [review]
@jordens Some questisons/comments on the design. Overall, I think we're pretty close to being able to publish this as something. Do we want to keep it in the stabilizer repo for now, or split it into (yet another) repo?

File: serial-settings/src/lib.rs
Comment (ryan-summers) [reply serial-settings/src/lib.rs:229]: Indeed! It should not be using `write_all()`, but instead just `write()`. However, I believe writing to `context` instead of `interface_mut()` is correct here, as writing to `interface_mut()` will indefinitely block if `WriteReady::write_ready()` is false, which will essentially stall the USB port and starve the idle task.

Comment (jordens) [reply serial-settings/src/lib.rs:229]: The problem is we don't write to `context` everywhere currently.

Comment (ryan-summers) [reply serial-settings/src/lib.rs:229]: Gotcha. I think this is a good reason why the interface should be explicitly owned by the `Context`, as it could then wrap the `Interface` in something that properly implements `core::fmt::Write`

File: serial-settings/src/lib.rs
Comment (jordens) [reply serial-settings/src/lib.rs:85]: The `Platform` equally accesses and uses the interface.
Ideally we'd want to merge `Platform` and `Context` into one (provided methods on the trait). But that seemed a bit hard when I tried.

File: serial-settings/src/lib.rs
Comment (jordens) [reply serial-settings/src/lib.rs:79]: In addition to the other points, it was less code to have Platform own settings and interface.

File: serial-settings/src/lib.rs
Comment (jordens) [reply serial-settings/src/lib.rs:85]: And also potentially get rid of the `Runner` as it doesn't seem to do much.

File: serial-settings/src/lib.rs
Comment (jordens) [reply serial-settings/src/lib.rs:226]: It doesn't seem appropriate to let this crate impl that trait.

Comment (ryan-summers) [reply serial-settings/src/lib.rs:217]: I fixed this in 2c391b669991052b5c450110bebcc7083ec989d4

File: serial-settings/src/lib.rs
Comment (ryan-summers) [reply serial-settings/src/lib.rs:82]: I've come around on `Platform` owning the interface, since indeed it seems like the platform really does need to use it quite often.

File: serial-settings/src/lib.rs
Comment (ryan-summers) [reply serial-settings/src/lib.rs]: I ended up removing `load()` entirely, since it didn't really serve much of a purpose (it was only being called once in the actual constructor of `Runner`). Instead, `Platform` is responsible for initializing its settings, which aligns with normal means of object construction.

**Comment by ryan-summers** [comment]
> @jordens Some questisons/comments on the design. Overall, I think we're pretty close to being able to publish this as something. Do we want to keep it in the stabilizer repo for now, or split it into (yet another) repo?

Two last comments/discussions left above, but otherwise I think things here are at a nice state. Ping on the question above as well.

File: serial-settings/src/lib.rs
Comment (jordens) [reply serial-settings/src/lib.rs:88]: To me "proper ownership" doesn't have clear and easy rules.

Settings changing outside this crate is something I can see coming up soon, e.g. also saving the application settings in flash and using that as the startup value (shadowed then later by mqtt where available and retained, I think I mentioned this somewhere else as well).

**Comment by jordens** [comment]
Let's split it once you want to use it on Booster. Is that now?

File: serial-settings/src/lib.rs
Comment (jordens) [inline serial-settings/src/lib.rs:313]: I still have an issue with this. Why does this crate need to make the write best effort? That seems like a thing for a downstream crate to decide.

**Review by jordens — CHANGES_REQUESTED** [review]

File: serial-settings/src/lib.rs
Comment (ryan-summers) [reply serial-settings/src/lib.rs:313]: Good point. I'll update it to not make assumptions about the writability of data.

Comment (jordens) [reply serial-settings/src/lib.rs:313]: I suspect that will also allow removing the dependency on eio traits from this crate.

Comment (ryan-summers) [reply serial-settings/src/lib.rs:313]: We need the `Read` component as well for the `process()` function to feed data to the menu interface. While we could replace the eio Write methods, it feels really odd to use half e-io and half core::fmt

Comment (jordens) [reply serial-settings/src/lib.rs:313]: agreed

Comment (jordens) [reply serial-settings/src/lib.rs:313]: But this does tie into the notion that `Runner` doesn't do much other than encapsulating `eio::Read`.

Comment (ryan-summers) [reply serial-settings/src/lib.rs:301]: Yeah, I recognize that. It largely comes down to whether we want this to be a "batteries included" approach or not. While I'm generally not a huge fan of making assumptions for the user, I think the `Runner` does eliminate a fair amount of boilerplate for the end user. Additionally, I added blanket-impls of the e-io traits for `usbd_serial::SerialPort`, so USB serial devices won't need any additional code to handle reading/writing data.

Ultimately, I'd say that we should keep the status quo and re-evaluate in the future if we want.

**Comment by ryan-summers** [comment]
> Let's split it once you want to use it on Booster. Is that now?

Meh, there's no huge rush to move it onto Booster yet, so maybe we just let it live here for the time being. We can break it out into a separate crate when we feel like it's mature enough.

**Pull request merged**
